File: sample_source_code/JAVA/src/test/java/com/embold/sample/ReportTest.java
package com.embold.sample;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

class ReportTest {

    private Shelter shelter;

    @BeforeEach
    void setUp() {
        shelter = new Shelter("Test Haven", 10);
        shelter.intake(new Dog("Rex",      3, "Labrador",  30.0));
        shelter.intake(new Dog("Buddy",    7, "Poodle",    15.0));
        shelter.intake(new Cat("Whiskers", 2, true,         4.5));
        shelter.intake(new Cat("Shadow",   5, false,        7.0));
    }

    @Test
    void generate_containsShelterName() {
        String report = new Report(shelter).generate();
        assertTrue(report.contains("Test Haven"));
    }

    @Test
    void generate_containsExpectedAnimalCounts() {
        String report = new Report(shelter).generate();
        assertTrue(report.contains("Dogs            : 2"));
        assertTrue(report.contains("Cats            : 2"));
        assertTrue(report.contains("Total animals   : 4"));
    }

    @Test
    void generate_containsAnimalDetails() {
        String report = new Report(shelter).generate();
        assertTrue(report.contains("Rex"));
        assertTrue(report.contains("Whiskers"));
    }

    @Test
    void generate_containsAvailableCapacity() {
        String report = new Report(shelter).generate();
        // 10 capacity - 4 animals = 6 available
        assertTrue(report.contains("Available cap   : 6"));
    }

    @Test
    void generate_emptyShelter_showsZeroCounts() {
        Shelter empty = new Shelter("Empty", 5);
        String report = new Report(empty).generate();
        assertTrue(report.contains("Total animals   : 0"));
        assertTrue(report.contains("Dogs            : 0"));
        assertTrue(report.contains("Cats            : 0"));
    }

    @Test
    void generate_returnsNonEmptyString() {
        String report = new Report(shelter).generate();
        assertNotNull(report);
        assertFalse(report.isBlank());
    }

    @Test
    void generate_calledTwice_returnsSameResult() {
        Report report = new Report(shelter);
        assertEquals(report.generate(), report.generate());
    }
}

File: sample_source_code/typescript/tests/report.test.ts
import { Report }  from '../src/Report';
import { Shelter } from '../src/Shelter';
import { Dog }     from '../src/Dog';
import { Cat }     from '../src/Cat';

function buildShelter(): Shelter {
  const shelter = new Shelter('Happy Paws', 20);
  shelter.intake(new Dog('Buddy',    3, 'Labrador',        28.5));
  shelter.intake(new Dog('Max',      7, 'German Shepherd', 33.0));
  shelter.intake(new Cat('Whiskers', 5, true,   4.2));
  shelter.intake(new Cat('Shadow',  10, false,  5.8));
  return shelter;
}

describe('Report.generate', () => {
  test('includes the shelter name', () => {
    expect(new Report(buildShelter()).generate()).toContain('Happy Paws');
  });
  test('includes total-animals count', () => {
    expect(new Report(buildShelter()).generate()).toContain('Total animals   : 4');
  });
  test('includes dog count', () => {
    expect(new Report(buildShelter()).generate()).toContain('Dogs            : 2');
  });
  test('includes cat count', () => {
    expect(new Report(buildShelter()).generate()).toContain('Cats            : 2');
  });
  test('includes available capacity', () => {
    expect(new Report(buildShelter()).generate()).toContain('Available cap   : 16');
  });
  test('lists animal names in details section', () => {
    const report = new Report(buildShelter()).generate();
    expect(report).toContain('Buddy');
    expect(report).toContain('Whiskers');
  });
  test('generate is idempotent', () => {
    const r = new Report(buildShelter());
    expect(r.generate()).toBe(r.generate());
  });
  test('empty shelter produces zero counts', () => {
    const report = new Report(new Shelter('Empty', 5)).generate();
    expect(report).toContain('Total animals   : 0');
    expect(report).toContain('Dogs            : 0');
    expect(report).toContain('Cats            : 0');
  });
});
